contrib/StochasticAlgo.py

```diff
@@ -0,0 +1,511 @@
+        
+
+    def update(self):
+        '''Single iteration'''
+        
+        self.x_update = self.objective_function.gradient(self.x)
+        
+        self.x_update *= -self.step_size
+        self.x += self.x_update
+
+    def update_objective(self):
+        self.loss.append(self.objective_function(self.x))
+
+# %%
+
+# Import libraries
+
+from cil.framework import  AcquisitionGeometry
+
+from cil.optimisation.functions import L2NormSquared, BlockFunction, MixedL21Norm, IndicatorBox
+from cil.optimisation.operators import GradientOperator, BlockOperator
+from cil.optimisation.algorithms import PDHG
+from cil.plugins.astra.operators import ProjectionOperator
+from cil.plugins.astra.processors import FBP
+from cil.plugins import TomoPhantom
+from cil.utilities.display import show2D, show_geometry
+from cil.utilities import noise
+
+import matplotlib.pyplot as plt
+
+import numpy as np
+# %%
+
+
+# Detectors
+N = 256
+detectors =  N
+
+# Angles
+angles = np.linspace(0,180,180, dtype='float32')
+
+# Setup acquisition geometry
+ag = AcquisitionGeometry.create_Parallel2D()\
+                        .set_angles(angles)\
+                        .set_panel(detectors, pixel_size=0.1)
+# Get image geometry
+ig = ag.get_ImageGeometry()
+
+# Get phantom
+phantom = TomoPhantom.get_ImageData(12, ig)
+
+# %%
+
+# Create projection operator using Astra-Toolbox. Available CPU/CPU
+A = ProjectionOperator(ig, ag, device = 'gpu')
+
+# Create an acqusition data (numerically)
+sino = A.direct(phantom)
+
+# Simulate Gaussian noise for the sinogram
+gaussian_var = 0.5
+gaussian_mean = 0
+
+n1 = np.random.normal(gaussian_mean, gaussian_var, size = ag.shape)
+                      
+noisy_sino = ag.allocate()
+noisy_sino.fill(n1 + sino.array)
+noisy_sino.array[noisy_sino.array<0]=0
+
+# Show numerical and noisy sinograms
+show2D([phantom, sino, noisy_sino], title = ['Ground Truth','Sinogram','Noisy Sinogram'], num_cols=3, cmap = 'inferno')
+
+
+# %%
+
+from cil.processors import Slicer
+from cil.optimisation.functions import LeastSquares
+
+# Define number of subsets
+n_subsets = 10
+
+# Initialize the lists containing the F_i's and A_i's
+f_subsets = []
+A_subsets = []
+
+# Define F_i's and A_i's
+for i in range(n_subsets):
+    # Total number of angles
+    n_angles = len(ag.angles)
+    # Divide the data into subsets
+    data_subset = Slicer(roi = {'angle' : (i,n_angles,n_subsets)})(noisy_sino)
+
+    # Define A_i and put into list 
+    ageom_subset = data_subset.geometry
+    Ai = ProjectionOperator(ig, ageom_subset)
+    A_subsets.append(Ai)
+
+    # Define F_i and put into list
+    fi = LeastSquares(Ai, b = data_subset)
+    f_subsets.append(fi)
+
+
+# Define F and K
+F = BlockFunction(*f_subsets)
+
+
+#%%
+
+F_saga = SAGA_Function(F) 
+
+initial = ig.allocate(0)
+gd = GradientDescent(initial=initial, objective_function=F_saga,
+                     step_size=0.0001, update_objective_interval=1, 
+                     max_iteration=1000)
```

Comment:
This comment is probably in the wrong place, but I guess we need to decide how do we want to compare different algorithms wrt number of subsets used. Or rather, should it be run with respect to total number of iterations or total number of epochs (both of which are independent of the number of subsets an algorithm uses, but the total number of iterations for a given number of epochs will depend on the number of subsets). I would argue that epoch counts should be a more relevant metric

Reply:
In SPDHG we use the number of iterations, not the number of epochs though, so the convention here should stay the same.

Reply:
We can add a counter for epochs as a method of each child of SubsetGradientFunction.

Reply:
agreed, it should be defined by each child (at the very least since the problem with epoch counts is svrg, and associated methods). so you mean that the code uses the number of iterations as input (not for comparison) - which i guess is a consequence of using the Algorithm class? (in general I completely agree that it would be better to be consistent)
the comparison on the spdhg github page (and the original paper) is wrt epoch counts, and eg the spdhg example https://github.com/mehrhardt/spdhg/blob/master/example_pet.py sets nepochs (to 300, in line 37) from which the number of iterations is computed (by just multiplying with num_subsets in l240) - having a functionality like that would be a bit more involved with svrg/katyusha, especially if you include different update frequencies. If we still want to have that functionality as in the example_pet code, one option might be to have a method that computes the number of iterations from the desired number of epochs?

Reply:
Yes, the code should use the number of iterations as input but the convergence comparison needs to be wrt epochs.
We could leave the computation of the number of iterations per epoch to the user as it's quite intricate (SAGA with and w/o warm-start, svrg etc...) or have an helper function for that.
For SPDHG we don't have such a helper function but we can certainly think about adding one. This would be a nice part of a more general "factory" where the user just provides the model and the desired algorithm and all the setting-up is done automatically and hidden. Notice also that for SPDHG the number of iterations per epoch is not always the number of subsets as in the example you referred to, in fact it depends on how the regularisation term is dealt with. In the "implicit" case, as in the example you referred to, the number of iterations per epoch is indeed the number of subsets, but on gets more complicated in the "explicit" case, see 
https://github.com/mehrhardt/spdhg_pet/blob/52e2c77b6327f97a0ae5b525de39ef50d105692c/python/map_tv.py#L243

Reply:
All of CIL `Algorithm`s do have a method `update` that is what happens during one iteration. There is no knowledge whatsoever in them about subset or epochs. Probably we could make a utility method in each of these algorithms that will convert epochs to iterations, if needed.

Notice, however that having moved the subsetting into the `Function` the algorithm will really go blind here. They could try to see if the objective function has a clue about it. The only algorithm that would know what to do is `SPDHG`.

Reply:
> Notice, however that having moved the subsetting into the Function the algorithm will really go blind here.

SVRG will need some special consideration in that case as we will not know if `update` is performing the "stochatic" or "full gradient" computations. Maybe an internal counter for number of sub-`Function` gradient calls have been made?

Reply:
of course, `SVRGFunction` will need to internal counter. No counter needed in GD. You could provide an `epoch()` function in the `SubsetFunction` (or whatever it is called now).

I guess in CIL you handle intermediate output with callbacks. They can have any additional counters that you like. I'd argue though that if someone sets up SVRG and want to plot in epochs, they can.